src/parser.ts

```diff
@@ -46,442 +35,536 @@ const archiveMarkerRegex = /^\*\*\*$/;
 
+    lines.push("");
+    lines.push("");
+    lines.push("");
+
     return lines.join("\n");
   }
 
-  return "";
-}
+  archiveToMd(archive: Item[]) {
+    if (archive.length) {
+      const lines: string[] = [archiveString, "", `## ${t("Archive")}`, ""];
 
-export function settingsToFrontmatter(settings: KanbanSettings): string {
-  return ["---", "", yaml.dump(settings), "---", "", ""].join("\n");
-}
+      archive.forEach((item) => {
+        lines.push(this.itemToMd(item));
+      });
 
-export function boardToMd(board: Board) {
-  const lanes = board.lanes.reduce((md, lane) => {
-    return md + laneToMd(lane);
-  }, "");
+      return lines.join("\n");
+    }
 
-  return (
-    settingsToFrontmatter(board.settings) + lanes + archiveToMd(board.archive)
-  );
-}
+    return "";
+  }
 
-export function mdToSettings(boardMd: string): KanbanSettings {
-  const match = boardMd.match(frontmatterRegEx);
+  settingsToFrontmatter(settings: KanbanSettings): string {
+    return ["---", "", yaml.dump(settings), "---", "", ""].join("\n");
+  }
+
+  boardToMd(board: Board) {
+    const lanes = board.lanes.reduce((md, lane) => {
+      return md + this.laneToMd(lane);
+    }, "");
 
-  if (match) {
-    return yaml.load(match[1].trim()) as KanbanSettings;
+    return (
+      this.settingsToFrontmatter(board.settings) + lanes + this.archiveToMd(board.archive)
+    );
   }
 
-  return { "kanban-plugin": "basic" };
-}
 
-export function mdToBoard(boardMd: string, view: KanbanView): Board {
-  const settings = mdToSettings(boardMd);
-  const lines = boardMd.replace(frontmatterRegEx, "").split(newLineRegex);
-  const lanes: Lane[] = [];
-  const archive: Item[] = [];
+  lastFrontMatter: string
+  lastSettings: KanbanSettings
+  lastGlobalSettings: KanbanSettings
+  lastItems: Map<string, Item[]> = new Map;
+  lastLanes: Map<string, Lane> = new Map;
+  fileCache: Map<TFile, FileMetadata> = new Map;
 
-  let haveSeenArchiveMarker = false;
+  // we use a string instead of a file, because a file changing path could change the meaning of links
+  lastParsedPath: string;
 
-  let currentLane: Lane | null = null;
+  invalidateFile(file: TFile) {
+    if (this.fileCache.has(file)) {
+      this.fileCache.delete(file);
+      return true;
+    }
+  }
 
-  lines.forEach((line) => {
-    if (archiveMarkerRegex.test(line)) {
-      haveSeenArchiveMarker = true;
+  mdToBoard(boardMd: string, filePath: string): Board {
+
+    /*
+    Steps:
+
+    1. Split front matter from board
+    2. If it's the same as the last, reuse the local settings calc
+    3. If the file path, global and local settings are all the same as before, reuse them
+    4. Otherwise, recalc settings and clear the cache
+
+    Issues:
+    * Should internal link be resolved from DOM instead of regex?
+    */
+
+    const [beforeFrontMatter, frontMatter, body] = boardMd.split(/^---\r?$\n?/m, 3);
+
+    if (beforeFrontMatter.trim()) throw new Error("Invalid Kanban file: problems parsing frontmatter ")
```

Comment:
Would this error surface to the user? If so, would you mind adding it to `src/lang/locale/en.ts` and making this update:

```suggestion
    if (beforeFrontMatter.trim()) throw new Error(t("Invalid Kanban file: problems parsing frontmatter"))
```

Reply:
Fixed and pushed.